pallets/gear/src/queue.rs

```diff
@@ -85,17 +85,7 @@ where
                 );
             }
 
-            let (dispatch, gas, _) = precharged_dispatch.into_parts();
-            return vec![
-                JournalNote::GasBurned {
-                    message_id: dispatch.id(),
-                    amount: gas.burned(),
```

Comment:
very important stuff and space for small refactoring: please make sure that

Kind::Init is impossible to be sent in uninitialised program - there is a panic
Kind::Handle is acceptable for this error-returning case
Kind::Reply is acceptable for run
Kind::HandleSignal is impossible - paste panic as well

Reply:
and speaking of those: it should be impossible to call system_reservation within the init, correct?
looks like this comment should be used as issue template

Reply:
> it should be impossible to call system_reservation within the init

It looks like signal is the only case for system reserve. But I don't see any reason to deny this behavior

Reply:
> Kind::Init is impossible to be sent in uninitialised program - there is a panic

`if` branches are introduced by @grishasobol

> Kind::Handle is acceptable for this error-returning case

`messages_to_uninitialized_program_wait`

> Kind::Reply is acceptable for run

`wake_messages_after_program_inited`

> Kind::HandleSignal is impossible - paste panic as well

Possible only pre-charging and actual execution

Reply:
> > it should be impossible to call system_reservation within the init
> 
> It looks like signal is the only case for system reserve. But I don't see any reason to deny this behavior

because handle signal relies on memory and program data which is wasted on failing init. I'll open an issue to consider wiping program after signal and unsuccessful init if deposit exists

Reply:
#3842 
I guess in case of opened issue its up to you whether to check it rn or not

Reply:
> because handle signal relies on memory and program data which is wasted on failing init. 

I mean, when we were designing system reservation there was assumption it will have more use-cases. So that's why I don't want to forbid it